feat!: add `as_polars_df` and `as_polars_lf` generic functions

eitsupi opened the pull request:
Related to #497
Add generic functions for creating DataFrames and LazyFrames.

The PR also includes breaking changes to some existing method arguments, but the probability of having code that depends on the order of these function arguments externally is low.

The default value of `n_rows` as `Inf` is based on dbplyr's `collect` function.

eitsupi commented on R/as_polars.R:
```diff
@@ -0,0 +1,109 @@
+
+
+#' @rdname as_polars_df
+#' @param n_rows Integer or `Inf`
```
TODO: this line should be removed

etiennebacher commented on R/lazyframe__lazy.R:
```diff
@@ -338,6 +338,7 @@ LazyFrame_set_optimization_toggle = function(
 #' @description `$collect()` performs the query on the LazyFrame. It returns a
 #' DataFrame
 #' @inheritParams LazyFrame_set_optimization_toggle
+#' @param ... Ignored.
```
I don't understand why you add this here. I get that the `as_polars()` methods must have these dots but nothing forces us to pass them to `$collect()` or `fetch()` in the method for `LazyFrame`

eitsupi replied:
The intent is to force these arguments to be specified by name.

etiennebacher commented on R/as_polars.R:
```diff
@@ -0,0 +1,109 @@
+#' @param ... Additional arguments passed to methods.
+#' @examples
+#' # Convert a data.frame to a polars LazyFrame
+#' lf = as_polars_df(mtcars)$lazy()
```
it would probably have more interest to show the example with the ArrowTable because for classic data.frames `pl$DataFrame()` works fine

etiennebacher commented on tests/testthat/test-as_polars.R:
```diff
@@ -0,0 +1,28 @@
+  tibble::tribble(
+    ~.test_name, ~x,
+    "data.frame", test_df,
+    "pldf", pl$DataFrame(test_df),
```
this case is not very useful since it's already used as the expectation

etiennebacher reviewed (commented):
Looks good, just a few comments, thanks!

etiennebacher reviewed (approved):
Thanks, can you also bump NEWS?

eitsupi commented:
I also added `as_polars_lf()` and Rust macro rewritten with `as_polars_*`.
This causes horizontal concat of multiple unnamed Series or R vectors to fail because they are now converted to DataFrame on the R side when horizontally concatting unnamed Series or R vectors.
However, the Python polars does not allow horizontal concat of Series in the first place, so this is not a problem.

etiennebacher commented on NEWS.md:
```diff
@@ -37,6 +37,8 @@
  is aimed for r-polars extensions, and will be kept stable as much as possible (#504).
 - New functions `pl$min_horizontal()`, `pl$max_horizontal()`, `pl$sum_horizontal()`,
   `pl$all_horizontal()`, `pl$any_horizontal()` (#508).
+- New generic functions `as_polars_df()` and `as_polars_lf` to create polars DataFrames
```
```suggestion
- New generic functions `as_polars_df()` and `as_polars_lf()` to create polars DataFrames and LazyFrames
```

eitsupi commented:
As a complete aside, I am surprised that Polars has no way to revert from LazyGroupBy to LazyFrame.
I am impressed that tidypolars seems to have implemented this well.......
(Personally, I think it is safer to rely on the `by` argument instead of using `dplyr::group_by`, as I had trouble dealing with groups even in the `arrow` package.)

etiennebacher commented:
> I am impressed that tidypolars seems to have implemented this well

Yeah that was annoying. The trick was to almost never use `$group_by()` but to store the groups in the attributes of the Data/LazyFrame and then internally check if there are some groups. It's one of the things I made in a specific PR if you want to take a look: https://github.com/etiennebacher/tidypolars/pull/22

eitsupi commented:
Looking at the source code, it is not possible to retrieve `opt_state` from LazyGroupBy, but it is almost the same as LazyFrame, so it seems possible to return to LazyFrame by recording `opt_state` on the r-polars side and then writing it back.

https://github.com/pola-rs/polars/blob/5240a35f42467227d9bca894e96ebfb31afe664c/crates/polars-lazy/src/frame/mod.rs#L62-L67
https://github.com/pola-rs/polars/blob/5240a35f42467227d9bca894e96ebfb31afe664c/crates/polars-lazy/src/frame/mod.rs#L1611-L1621

I don't know if it is worth doing.

eitsupi commented on R/as_polars.R:
```diff
@@ -0,0 +1,156 @@
+
+#' @rdname as_polars_df
+#' @export
+as_polars_df.GroupBy = function(x, ...) {
+  x$to_data_frame()
+}
```
Sorry, this was incorrect.......
Not polars DataFrame, but R data.frame.
I will open a follow-up PR.

eitsupi replied:
#522
